Source/Core/Core/IOS/ES/ES.cpp

```diff
@@ -324,63 +330,84 @@ void ES::DoState(PointerWrap& p)
 
 ReturnCode ES::Open(const OpenRequest& request)
 {
-  m_caller_uid = request.uid;
-  m_caller_gid = request.gid;
+  auto context =
+      std::find_if(m_contexts.begin(), m_contexts.end(), [](const auto& c) { return !c.active; });
+  if (context == m_contexts.end())
+    return ES_FD_EXHAUSTED;
+
+  context->active = true;
+  context->uid = request.uid;
+  context->gid = request.gid;
+  context->ipc_fd = request.fd;
   return Device::Open(request);
 }
 
-void ES::Close()
+ReturnCode ES::Close(u32 fd)
 {
-  // XXX: does IOS really clear the content access map here?
+  auto context = std::find_if(m_contexts.begin(), m_contexts.end(),
```

Comment:
This and the `find_if` in IOCtlV() appear to do the same thing, just with different captures. Can it not be moved to it's own `<whatever  the iterator type is> FindFirstActiveContext(u32 fd)` function? (using whatever name is more appropriate)?